src/layers/regularizers/entrywise_batch_normalization.cpp

```diff
@@ -0,0 +1,423 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/layers/regularizers/entrywise_batch_normalization.hpp"
+
+namespace lbann {
+
+namespace {
+
+// Block size for loops
+// Note: x86 cache lines are 64B
+constexpr El::Int _bsize = 64 / sizeof(DataType);
+constexpr El::Int bsize = _bsize > 1 ? _bsize : 1;
+
+/**
+ *  mean = sum(x_i) / n
+ *
+ *  var = ( sum(x_i^2)/n - mean^2 ) * n/(n-1)
+ */
+void compute_batch_statistics(lbann_comm& comm,
+                              DataType decay,
+                              const AbsDistMat& input,
+                              AbsDistMat& batch_statistics,
+                              AbsDistMat& running_mean,
+                              AbsDistMat& running_var) {
+
+  // Local matrices
+  const auto& local_input = dynamic_cast<const CPUMat&>(input.LockedMatrix());
+  auto& local_batch_statistics = dynamic_cast<CPUMat&>(batch_statistics.Matrix());
+  auto local_batch_mean = El::View(local_batch_statistics, El::ALL, El::IR(0));
+  auto local_batch_var = El::View(local_batch_statistics, El::ALL, El::IR(1));
+  auto& local_running_mean = dynamic_cast<CPUMat&>(running_mean.Matrix());
+  auto& local_running_var = dynamic_cast<CPUMat&>(running_var.Matrix());
+
+  // Dimensions
+  const El::Int local_height = local_input.Height();
+  const El::Int local_width = local_input.Width();
+
+  // Compute local sums
+  El::Zero(batch_statistics);
+  LBANN_OMP_PARALLEL_FOR
+  for (El::Int row_start = 0; row_start < local_height; row_start += bsize) {
+    const El::Int row_end = std::min(row_start + bsize, local_height);
+    const El::Int col_start = 0;
+    const El::Int col_end = local_width;
+    for (El::Int col = col_start; col < col_end; ++col) {
+      for (El::Int row = row_start; row < row_end; ++row) {
+        const auto& x = local_input(row, col);
+        local_batch_mean(row, 0) += x;
+        local_batch_var(row, 0) += x * x;
+      }
+    }
+  }
+
+  // Accumulate sums between processes
+  /// @todo Local statistics
+  /// @todo Arbitrary group sizes
+  comm.allreduce(batch_statistics,
+                 batch_statistics.RedundantComm(),
+                 El::mpi::SUM);
+  const size_t statistics_count = input.Width();
+
+  // Compute mini-batch statistics from sums
+  if (statistics_count <= 1) {
+    // local_mean already has correct values
+    El::Fill(local_batch_var, DataType{1});
```

Comment:
This case occurs when the mini-batch size being aggregated over is 1 or less, right? We may want to throw an error instead, since the variance isn't well-defined then.

Reply:
I think that's too harsh, since the user will typically have little control on the size of the last mini-batch.